### PR TITLE
Merge `KeyEventExtra` into `KeyEvent`

### PR DESCRIPTION
**Pull request by madsmtm**
To make the `text_with_all_modifiers` and `key_without_modifiers` fields easier to use, and to allow constructing `KeyEvent` in user test code. This also removes the need for `KeyEventExtModifierSupplement`.

Fixes https://github.com/rust-windowing/winit/issues/3606 (which [is desired by Xilem/Masonry](https://xi.zulipchat.com/#narrow/channel/317477-masonry/topic/Important.20Winit.20issues.20for.20Masonry)).

I went with adding the fields on all platforms, and just defining their behaviour as the same as `text`/`logical_key` on the unsupported platforms. This is similar to how we have all `WindowEvent`s available on all the platforms, but not all are emitted by all of them, so I felt this was a good trade-off.

- [ ] Tested on all platforms changed
- [x] Added an entry to the `changelog` module if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality

### CODE REVIEW

File: src/event.rs [hunk 1]
```diff
@@ -862,13 +855,33 @@ pub struct KeyEvent {
...
+    /// layout. This value cannot be `Dead`.
+    ///
+    /// ## Platform-specific
     ///
-    /// On Android, iOS, Redox and Web, this type is a no-op.
-    pub(crate) platform_specific: platform_impl::KeyEventExtra,
+    /// - **Android:** Unimplemented, this field is always the same value as `logical_key`.
+    /// - **iOS:** Unimplemented, this field is always the same value as `logical_key`.
+    /// - **Web:** Unsupported, this field is always the same value as `logical_key`.
+    pub key_without_modifiers: keyboard::Key,
```
Comment (madsmtm) [inline src/event.rs:884]: Another option would be to make this an `Option<keyboard::Key>`, I'm not familiar enough with it to say what would be best.

File: src/event.rs [hunk 2]
```diff
@@ -862,13 +855,33 @@ pub struct KeyEvent {
...
-    /// Platform-specific key event information.
+    /// Identical to [`text`][Self::text] but this is affected by <kbd>Ctrl</kbd>.
+    ///
+    /// For example, pressing <kbd>Ctrl</kbd>+<kbd>a</kbd> produces `Some("\x01")`.
+    ///
+    /// ## Platform-specific
+    ///
+    /// - **Android:** Unimplemented, this field is always the same value as `text`.
+    /// - **iOS:** Unimplemented, this field is always the same value as `text`.
+    /// - **Web:** Unsupported, this field is always the same value as `text`.
```
Comment (madsmtm) [inline src/event.rs:866]: Another option would be to always set `None` here on these platforms, but I felt this choice was more consistent with `key_without_modifiers` (unless we change that to be an `Option`).

**Review by kchibisov — CHANGES_REQUESTED** [review]
Could we have it the same way as before in a struct on a `KeyboardEvent` which you can not match, but can e.g. default init or set separately?

I don't really want to have an option to _match_ on those, since they are not cross platform. Other than that removing the extension is fine and it's not that big of a breaking change.

**Comment by madsmtm** [comment]
> Could we have it the same way as before in a struct on a `KeyboardEvent` which you can not match, but can e.g. default init or set separately?

Hmm, I can, but I kinda feel like that'd just be unnecessary indirection? And I guess it we really want users to be able to test their crates, they'll need the ability to set the fields in this extra struct anyhow?

> I don't really want to have an option to _match_ on those, since they are not cross platform. Other than that removing the extension is fine and it's not that big of a breaking change.

Am I understanding correctly that your worry is that users may be more aware of a method not being cross-platform, as opposed to a field being so?

**Comment by kchibisov** [comment]
Yes, so it's an `Option` and you should account for it explicitly. If it's not an option you will kind of assume that it works, where on e.g. some it'll always fail. You can just default initialize it, I don't see an issue with it. Yes you won't be able to test the values you have in `text_with_all_modifiers`, but you write the keys yourself, so can write whatever in regular fields, so it doesn't really matter.

**Comment by madsmtm** [comment]
I've been thinking about this again, and I don't think there's a good reason why these are treated differently from the other `KeyEvent` fields.

Remember that the fact that `text_with_all_modifiers` and `key_without_modifiers` doesn't work on Android/iOS is merely an artifact of the current implementation, they very much _could_ work there. The only platform where the situation may be a bit different is Web, but even there, we could probably implement it? At least `text_with_all_modifiers` is "just" a matter of mapping [`key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) + check [`ctrlKey`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/ctrlKey) according to [a basic ASCII table](https://ss64.com/ascii.html), no?

Basically, I'm arguing that these APIs are _supportable_ on all platforms, and hence should be available everywhere too; just like `WindowEvent::Moved` is available everywhere, even though it's not yet implemented on many platforms.

**Review by kchibisov — APPROVED** [review]
I think it's fine, I don't like this API at all, and would just deprecate it incrementally in the future by having `KeyboardHanlingV1` and `V2` where `V1` is implemented via `V2` and apps supporting `V2` will get something more reliable.
